src/Temporalio/Workflows/WorkflowDefinition.cs

```diff
@@ -427,6 +428,35 @@ public static WorkflowDefinition Create(
                 }
             }
 
+            // Verify that registered names to not use our reserved prefix
+            if (name != null && name.StartsWith(Constants.ReservedNamePrefix))
+            {
+                errs.Add($"Workflow name {name} cannot start with {Constants.ReservedNamePrefix}");
+            }
+            void CheckNamesForPrefix(IEnumerable<string> keys, string type)
+            {
+                foreach (var key in keys)
+                {
+                    if (key.StartsWith(Constants.ReservedNamePrefix))
+                    {
+                        errs.Add($"{type} handler name {key} cannot start with {Constants.ReservedNamePrefix}");
+                    }
+                    else if (type == "Query")
+                    {
+                        foreach (var reservedQ in Workflow.ReservedQueryHandlerPrefixes)
+                        {
+                            if (key.StartsWith(reservedQ))
+                            {
+                                errs.Add($"{type} handler name {key} cannot start with {reservedQ}");
+                            }
+                        }
+                    }
+                }
+            }
+            CheckNamesForPrefix(signals.Keys, "Signal");
+            CheckNamesForPrefix(queries.Keys, "Query");
+            CheckNamesForPrefix(updates.Keys, "Update");
```

Comment:
I think you can just check signal name prefix in signal definition, update in update definition, and query in query definition. It is clearer to keep validation with the definition.

Reply:
The downside of this is now you need a specific bypassing constructor -- see my last commit